File: frontend/lib/services/message_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config.dart';
import '../models/message.dart';
import 'auth_service.dart';

class MessageService {
  final AuthService _authService = AuthService();

  Future<Message> sendMessage({
    required int matchId,
    required String content,
  }) async {
    final headers = await _authService.getAuthHeaders();
    final response = await http.post(
      Uri.parse('$API_URL/api/messages'),
      headers: headers,
      body: jsonEncode({
        'matchId': matchId,
        'content': content,
      }),
    );

    if (response.statusCode == 200) {
      return Message.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to send message');
    }
  }

  Future<List<Message>> getConversation(int matchId) async {
    final headers = await _authService.getAuthHeaders();
    final response = await http.get(
      Uri.parse('$API_URL/api/messages/match/$matchId'),
      headers: headers,
    );

    if (response.statusCode == 200) {
      final List<dynamic> data = jsonDecode(response.body);
      return data.map((json) => Message.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load conversation');
    }
  }

  Future<void> markAsRead(int messageId) async {
    final headers = await _authService.getAuthHeaders();
    await http.put(
      Uri.parse('$API_URL/api/messages/$messageId/read'),
      headers: headers,
    );
  }

  Future<int> getUnreadCount() async {
    final headers = await _authService.getAuthHeaders();
    final response = await http.get(
      Uri.parse('$API_URL/api/messages/unread-count'),
      headers: headers,
    );

    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return data['count'] ?? 0;
    } else {
      return 0;
    }
  }
}

File: frontend_react/src/services/message.service.js
import api from '../api/axios';
import { normalizeMessage } from '../utils/normalizers';

const getCurrentUserId = () => {
    try {
        const raw = localStorage.getItem('user');
        if (!raw) return null;
        const user = JSON.parse(raw);
        return typeof user?.id === 'number' ? user.id : Number(user?.id);
    } catch (_e) {
        return null;
    }
};

const formatTime = (isoString) => {
    if (!isoString) return '';
    const date = new Date(isoString);
    if (Number.isNaN(date.getTime())) return '';
    return date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
};

export const messageService = {
    getMessages: async (matchId) => {
        const response = await api.get(`/messages/match/${matchId}`);
        const list = Array.isArray(response.data) ? response.data : [];
        const currentUserId = getCurrentUserId();

        return list.map((m) => {
            const normalized = normalizeMessage(m, currentUserId);
            if (!normalized) return null;
            return {
                id: normalized.id,
                text: normalized.text,
                matchId: normalized.matchId ?? matchId,
                isMe: normalized.isMe,
                time: formatTime(normalized.sentAt),
                status: normalized.isRead ? 'read' : 'sent',
                sentAt: normalized.sentAt,
                senderId: normalized.senderId,
                recipientId: normalized.recipientId,
                isRead: normalized.isRead,
                messageType: normalized.messageType ?? 'TEXT',
                referenceId: normalized.referenceId,
            };
        }).filter(Boolean);
    },

    sendMessage: async (matchId, text) => {
        const response = await api.post('/messages', { matchId, content: text });
        const currentUserId = getCurrentUserId();
        const normalized = normalizeMessage(response.data, currentUserId);
        if (!normalized) return null;
        return {
            id: normalized.id,
            text: normalized.text,
            matchId: normalized.matchId ?? matchId,
            isMe: normalized.isMe,
            time: formatTime(normalized.sentAt),
            status: normalized.isRead ? 'read' : 'sent',
            sentAt: normalized.sentAt,
            senderId: normalized.senderId,
            recipientId: normalized.recipientId,
            isRead: normalized.isRead,
            messageType: normalized.messageType ?? 'TEXT',
            referenceId: normalized.referenceId,
        };
    },

    markAsRead: async (messageId) => {
        await api.put(`/messages/${messageId}/read`);
        return true;
    },

    getUnreadCount: async () => {
        const response = await api.get('/messages/unread-count');
        return response.data?.count ?? 0;
    },

    reportMessage: async (messageId, reason) => {
        console.log(`Reporting message ${messageId} for: ${reason}`);
        return { success: true };
    },

    decryptMessage: async (messageId) => {
        console.log(`Decrypting/Revealing message ${messageId}`);
        return { success: true };
    }
};

